### PR TITLE
Add warning when training is not set explicitly in base Module class

### PR DESCRIPTION
**Pull request by ListIndexOutOfRange**
Currently, the training attribute of the `Module` base class is set to `True` by default. This can lead to unwanted and hard to catch side-effects.
For instance, when inferring using a `model` with `BatchNorm`, forgetting to call `model.eval()` could lead to undesired behavior (running stats being silently updated).
However, throwing an error seems too strict as it would break many existing code.
This PR instead proposes a very basic warning mechanism:
- the `training` attribute is set to `None` by default (instead of `True`)
- when the `forward` method is called, the base `Module` class checks if `self.training` is `None` and if so, raises a warning.

### CODE REVIEW

**Comment by albanD** [comment]
Hi!
Thanks for sending this PR, we usually prefer to discuss this kind of proposals on an issue before you spend time sending a PR.

This change will be challenging I'm afraid. In particular, changing the default state like that is very much Backward Compatibility (BC) breaking. Also I think such a warning would be way too verbose as a lot of users don't set train/eval explicitly in many cases where it is not needed and this warning will be quite spammy for them.

**Comment by ListIndexOutOfRange** [comment]
Hi ! 
Thanks for your quick reply !

Yes sorry about the too quick PR, I thought it didn't matter since the PR was so small. 

Of course the backward compatibility is a big issue. I still think forcing explicit call to .train() or .eval() could be beneficial.  I think the vast majority of existing training codes / frameworks already explicitly call .train() and .eval(). On the other hand, many new pytorch users don't think about it and encounter issues that could be easily mitigated this way. 
As for the verbosity, I personally would like pytorch to warn a bit more in this case. 

Consider for instance:

```python
batch_inputs = torch.randn(2, 3, 16, 16)
bn_train = torch.nn.BatchNorm2d(3).train()
bn_eval = torch.nn.BatchNorm2d(3).eval()
batch_outputs_train = bn_train(batch_inputs)
single_output_train = bn_train(batch_inputs[0].unsqueeze(0))
batch_outputs_eval = bn_eval(batch_inputs)
single_output_eval = bn_eval(batch_inputs[0].unsqueeze(0))
print((batch_outputs_train[0] == single_output_train).all())  # False
print((batch_outputs_eval[0] == single_output_eval).all())   # True 
```

I think this kind of behavior is challenging for new users (especially when hidden inside a external call to an existing model), whereas a small warning would clarify things immediately. 

Do you think the idea should be entirely discarded ? Or maybe there is a better way to warn the user ? Perhaps we could set training to True by default and catch if the attribute is modified so that the user would still get the warning but there would be no BC issue.

**Comment by albanD** [comment]
Thanks for the detailed answer. I want to get some more opinions on this so will take a couple days to answer!

**Comment by jbschlosser** [comment]
Discussed internally: we would prefer not to spam users with a warning requiring a `train()` / `eval()` call, at least by default, as it appears there are lots of instances where neither is called at the moment.

Perhaps there could be an opt-in mode enabling these types of guards / warnings for common new user footguns? These types of requests come through relatively frequently and we generally avoid them in favor of existing code / more experienced users. There's a reason for this tradeoff, but I don't want to discount new user difficulties.
